compiler/rustc_mir_transform/src/instcombine.rs

```diff
@@ -0,0 +1,405 @@
+) -> Option<StatementKind<'tcx>> {
+    let Rvalue::Use(temp_operand) = temp_rvalue else { return None; };
+    let Rvalue::Use(final_operand) = final_rvalue else { return None; };
+
+    // Both the operands needs to be places (is it even possible for one to be a Constant?)
+    let first_place = temp_operand.place()?;
+    let second_place = final_operand.place()?;
+
+    // If we are assigning into a place expression, that would be something like
+    // _3.b = _2.b;
+    // Which is complicated. Just don't optimize that at all for now.
+    if !temp_place.projection.is_empty() {
+        return None;
+    }
+
+    // Derefs must come first if at all
+    // If merging these assignments would break that rule, bummer. Bail.
+    if first_place.projection.len() > 0
+        && second_place.projection.get(0) == Some(&ProjectionElem::Deref)
+    {
+        return None;
+    }
+
+    // See: rust-lang/rust#11518
+    // If the temporary has a niche but the final does not, doing this optimization will destroy
+    // the niche information.
+    // This check is _extremely_ cautious, we only do this optimization if we are absolutely
+    // certain that the temporary does not have a niche.
+    // Note that if the second assignment does not add any projections, layout can't change, so we
+    // don't need this check.
+    if !second_place.projection.is_empty() {
+        let temporary_ty = temp_place.ty(cx.local_decls, cx.tcx).ty;
+        let Ok(layout) = cx.tcx.layout_of(cx.param_env.and(temporary_ty)) else {
+            return None;
+        };
+        if layout.layout.largest_niche().is_some() {
+            return None;
+        }
+    }
+
+    let new_place = first_place.project_deeper(&second_place.projection[..], cx.tcx);
+
+    if new_place == *final_place {
+        Some(StatementKind::Nop)
+    } else if final_operand.is_move() {
+        // If the Operand in the second statement is Move, the Place it refers to may be unsized,
+        // which would be wrong to copy, so we need to emit Operand::Move.
+        Some(StatementKind::Assign(Box::new((*final_place, Rvalue::Use(Operand::Move(new_place))))))
+    } else {
+        Some(StatementKind::Assign(Box::new((*final_place, Rvalue::Use(Operand::Copy(new_place))))))
+    }
+}
+
+// Combine:
+// _2 = _1 as *const T;
+// _3 = _2 as *const U;
+// Into:
+// _3 = _1 as *const U;
+fn combine_ptr_ptr_cast<'tcx, 'a>(
+    _cx: &Context<'tcx, 'a>,
+    temp_place: &Place<'tcx>,
+    temp_rvalue: &Rvalue<'tcx>,
+    final_place: &Place<'tcx>,
+    final_rvalue: &Rvalue<'tcx>,
+) -> Option<StatementKind<'tcx>> {
+    if !temp_place.projection.is_empty() {
+        return None;
+    }
+    let Rvalue::Cast(CastKind::PtrToPtr | CastKind::Pointer(PointerCast::MutToConstPointer), temp_operand, _temp_ty) = temp_rvalue else { return None; };
+    let Rvalue::Cast(CastKind::PtrToPtr | CastKind::Pointer(PointerCast::MutToConstPointer), _final_operand, final_ty) = final_rvalue else { return None; };
```

Comment:
Likewise, should we test that `temp_place.local == _final_operand`?
What about:
```rust
_2 = _1 as *const T;
_3 = (*_2).a as *const U;
```